File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseWeeklyOverviewAdapter.java
package br.com.jonathanzanella.myexpenses.expense;

import android.content.Intent;
import android.os.AsyncTask;
import android.support.annotation.UiThread;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;

import br.com.jonathanzanella.myexpenses.R;
import br.com.jonathanzanella.myexpenses.card.CreditCardInvoiceActivity;
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable;
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper;
import butterknife.BindView;
import butterknife.ButterKnife;

public class ExpenseWeeklyOverviewAdapter extends RecyclerView.Adapter<ExpenseWeeklyOverviewAdapter.ViewHolder> {
	private static final SimpleDateFormat SIMPLE_DATE_FORMAT = new SimpleDateFormat("dd", Locale.getDefault());
	protected List<Expense> expenses;
	private int totalValue;

	public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
		@BindView(R.id.row_weekly_overview_expense_name)
		TextView name;
		@BindView(R.id.row_weekly_overview_expense_date)
		TextView date;
		@BindView(R.id.row_weekly_overview_expense_income)
		TextView income;
		@BindView(R.id.row_weekly_overview_expense_source)
		TextView source;

		public ViewHolder(View itemView) {
			super(itemView);

			ButterKnife.bind(this, itemView);

			itemView.setOnClickListener(this);
		}

		@UiThread
		public void setData(final Expense expense) {
			name.setText(expense.getName());
			synchronized (this) {
				date.setText(SIMPLE_DATE_FORMAT.format(expense.getDate().toDate()));
			}
			income.setText(CurrencyHelper.format(expense.getValueToShowInOverview()));

			new AsyncTask<Void, Void, Chargeable>() {

				@Override
				protected Chargeable doInBackground(Void... voids) {
					return expense.getChargeableFromCache();
				}

				@Override
				protected void onPostExecute(Chargeable chargeable) {
					super.onPostExecute(chargeable);
					source.setText(chargeable.getName());
				}
			}.execute();
		}

		@Override
		public void onClick(View v) {
			Expense expense = getExpense(getAdapterPosition());
			if(expense != null) {
				if(expense.getCreditCard() != null) {
					Intent i = new Intent(itemView.getContext(), CreditCardInvoiceActivity.class);
					i.putExtra(CreditCardInvoiceActivity.KEY_CREDIT_CARD_UUID, expense.getCreditCard().getUuid());
					i.putExtra(CreditCardInvoiceActivity.KEY_INIT_DATE, expense.getDate());
					itemView.getContext().startActivity(i);
				} else {
					Intent i = new Intent(itemView.getContext(), ShowExpenseActivity.class);
					i.putExtra(ShowExpenseActivity.KEY_EXPENSE_UUID, expense.getUuid());
					itemView.getContext().startActivity(i);
				}
			}
		}
	}

	@Override
	public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
		View v = LayoutInflater.from(parent.getContext()).inflate(R.layout.row_weekly_overview_expense, parent, false);
		return new ViewHolder(v);
	}

	@Override
	public void onBindViewHolder(ViewHolder holder, int position) {
		holder.setData(expenses.get(position));
	}

	@Override
	public int getItemCount() {
		return expenses != null ? expenses.size() : 0;
	}

	public void setExpenses(List<Expense> expenses) {
		this.expenses = expenses;
		totalValue = 0;

		for (Expense expense : expenses)
			totalValue += expense.getValueToShowInOverview();
	}

	private Expense getExpense(int position) {
		return expenses.get(position);
	}

	public List<Expense> getExpenses() {
		return expenses;
	}

	public int getTotalValue() {
		return totalValue;
	}
}

File: app/src/main/java/br/com/jonathanzanella/myexpenses/expense/ExpenseWeeklyOverviewAdapter.kt
package br.com.jonathanzanella.myexpenses.expense

import android.content.Intent
import android.os.AsyncTask
import android.support.annotation.UiThread
import android.support.v7.widget.RecyclerView
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import br.com.jonathanzanella.myexpenses.R
import br.com.jonathanzanella.myexpenses.card.CreditCardInvoiceActivity
import br.com.jonathanzanella.myexpenses.chargeable.Chargeable
import br.com.jonathanzanella.myexpenses.helpers.CurrencyHelper
import br.com.jonathanzanella.myexpenses.views.anko.applyTemplateViewStyles
import br.com.jonathanzanella.myexpenses.views.anko.resumeRowCell
import org.jetbrains.anko.*
import java.text.SimpleDateFormat
import java.util.Locale
import kotlin.collections.ArrayList

class ExpenseWeeklyOverviewAdapter : RecyclerView.Adapter<ExpenseWeeklyOverviewAdapter.ViewHolder>() {
    private var expenses: List<Expense> = ArrayList()
    var totalValue: Int = 0
        private set

    inner class ViewHolder internal constructor(itemView: View, val ui : ExpenseWeeklyOverviewUI) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
        init {
            itemView.setOnClickListener(this)
        }

        @UiThread
        fun setData(expense: Expense) {
            ui.name.text = expense.name
            synchronized(this) {
                ui.date.text = SIMPLE_DATE_FORMAT.format(expense.date.toDate())
            }
            ui.income.text = CurrencyHelper.format(expense.valueToShowInOverview)

            object : AsyncTask<Void, Void, Chargeable>() {

                override fun doInBackground(vararg voids: Void): Chargeable {
                    return expense.chargeableFromCache
                }

                override fun onPostExecute(chargeable: Chargeable) {
                    super.onPostExecute(chargeable)
                    ui.source.text = chargeable.name
                }
            }.execute()
        }

        override fun onClick(v: View) {
            val expense = getExpense(adapterPosition)
            if (expense != null) {
                if (expense.creditCard != null) {
                    val i = Intent(itemView.context, CreditCardInvoiceActivity::class.java)
                    i.putExtra(CreditCardInvoiceActivity.KEY_CREDIT_CARD_UUID, expense.creditCard.uuid)
                    i.putExtra(CreditCardInvoiceActivity.KEY_INIT_DATE, expense.date)
                    itemView.context.startActivity(i)
                } else {
                    val i = Intent(itemView.context, ShowExpenseActivity::class.java)
                    i.putExtra(ShowExpenseActivity.KEY_EXPENSE_UUID, expense.uuid)
                    itemView.context.startActivity(i)
                }
            }
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val ui = ExpenseWeeklyOverviewUI()
        return ViewHolder(ui.createView(AnkoContext.create(parent.context, parent)), ui)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.setData(expenses[position])
    }

    override fun getItemCount(): Int {
        return expenses.size
    }

    fun setExpenses(expenses: List<Expense>) {
        this.expenses = expenses
        totalValue = expenses.sumBy { it.valueToShowInOverview }
    }

    private fun getExpense(position: Int): Expense? {
        return expenses[position]
    }

    companion object {
        private val SIMPLE_DATE_FORMAT = SimpleDateFormat("dd", Locale.getDefault())
    }
}

internal class ExpenseWeeklyOverviewUI : AnkoComponent<ViewGroup> {
    lateinit var date: TextView
    lateinit var source: TextView
    lateinit var name: TextView
    lateinit var income: TextView

    override fun createView(ui: AnkoContext<ViewGroup>) = with(ui) {
        verticalLayout {
            resumeRowCell {
                date = textView { id = R.id.row_weekly_overview_expense_date }
                        .lparams { marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing) }
                source = textView { id = R.id.row_weekly_overview_expense_source }
                        .lparams(width = resources.getDimensionPixelSize(R.dimen.expense_weekly_overview_source_width)) {
                            marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing)
                        }
                name = textView { id = R.id.row_weekly_overview_expense_name }
                        .lparams(width = 0, weight = 1f) {
                            marginEnd = resources.getDimensionPixelSize(R.dimen.min_spacing)
                        }
                income = textView { id = R.id.row_weekly_overview_expense_income }
            }
        }.applyRecursively(::applyTemplateViewStyles)
    }
}
